src/HashedInterp.hs

```diff
@@ -773,42 +773,60 @@ instance Evaluable Three C (Array (Int, Int, Int) (Complex Double)) where
                     _ -> error "expression structure Three C is wrong"
         | otherwise = error "Three C but shape is not [size1, size2, size3] ??"
 
--- |
+-- | Calculate the real shift amount. Based on the  position in the list, size of the shift and length fo the range.
 --
 rtt :: Int -> Int -> Int
 rtt pos size = (pos + size) `mod` size
 
--- |
+-- | Calculate the real shift amount. Based on the  position in the list, size of the shift and length fo the range.
 --
-rotate1D :: Int -> Int -> Array Int a -> Array Int a
+rtt1 :: Int -- ^ Input : Position in the list
+  -> Int  -- ^ Input : Real size of the array
+  -> Int  -- ^ Input : Amount of the shift
+  -> Int -- ^ Output : Real shift amount
+rtt1 pos size amount = (pos + (amount `mod` size)) `mod` size
+
+
+
+-- | One dimension rotation
+--
+rotate1D ::
+  Int -- ^ Input : Size of the array
+  -> Int -- ^ Input : amount of shift
+  -> Array Int a -- ^ Input : Input array
+  -> Array Int a -- ^ Output : Output, rotated array
 rotate1D size amount arr =
-    listArray (0, size) [arr ! rtt (i - amount) size | i <- [0 .. size - 1]]
+    listArray (0, size-1) [arr ! rtt1 i size  (-amount) | i <- [0 .. size - 1]]
```

Comment:
OK, I stand corrected. But this still doing left shift. Basic idea is to do the right shift if the user's input is a positive integer.

Reply:
I don't get what you say. We don't care about the direction of the rotating in `rtt (i - amount) size`. If the amount is positive, it is rotating to the right, and left otherwise. But the new positition at `i` always corresponds to the position `i - amount` in the original array (which may fall out of the range and we need to bring it back).